### DISCUSSION TITLE
View EVENT_BEFORE_RENDER cannot change $params - possible feature request

### DISCUSSION THREAD

**Discussion by papppeter** [yiisoft/yii2#19273] [General]
### What steps will reproduce the problem?

Subscribe to View::class, View::EVENT_BEFORE_RENDER

```
Event::on(View::class, View::EVENT_BEFORE_RENDER, function (ViewEvent $event) {
            /** @var View $sender */
            $sender = $event->sender;
            $event->params['context'] = $event->params['context']??$sender->context;
        });
```

try to add an extra parameter

### What is the expected result?

to be able to adjust params

### What do you get instead?

params is not extended because it is not used as a reference

### Suggestion

it would be nice if i could add a flag to ViewEvent if i would like to use params, and viewFile as a reference, or if it should overwrite the original values.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.45?
| PHP version      | 7.24
| Operating system | Windows

**Comment by WinterSilence** [comment]
~~~php
$view->on(View::EVENT_BEFORE_RENDER, function (ViewEvent $event) {
    $view = $event->sender;
    $ext = pathinfo($event->viewFile, PATHINFO_EXTENSION);
    if (isset($view->renderers[$ext])) {
        if (!is_object($view->renderers[$ext])) {
            $view->renderers[$ext] = Yii::createObject($view->renderers[$ext]);
        }
        /** @var ViewRenderer $renderer */
        $renderer = $view->renderers[$ext];
        $event->output = $renderer->render($view, $event->viewFile, $event->params);
    } else {
        $event->output = $view->renderPhpFile($event->viewFile, $event->params);
    }
    $view->afterRender($event->viewFile, $event->params, $event->output);
    $event->isValid = false;
});
~~~
or
~~~php
return [
  ...
  'components' => [
    'view' => [
       'class' => View::class,
       'on ' . View::EVENT_BEFORE_RENDER => function (ViewEvent $event) {
          ...
       }
    ],
  ],
];
~~~

**Comment by WinterSilence** [comment]
easy way: is set `$params` as `ArrayObject()`:
~~~php
$view->renderFile($viewFile, new ArrayObject($params), $context);
~~~

**Comment by papppeter** [comment]
Both looks promising. i could do it locally, but do you think it would be framework improvement?

**Comment by WinterSilence** [comment]
@papppeter "easy way" is popular native solution, something like this using in reactive JS to detect variable changes. Also, you can do it by render behaviour or extend renderer and override  `ViewRenderer::render()`. The approach is chosen depending on current case/goal.

**Comment by papppeter** [comment]
@WinterSilence thanks! i will try it out. 

Should i close this issue?

**Comment by WinterSilence** [comment]
@papppeter as you wish, I'm not repo owner (: Do you see [discussions](https://github.com/yiisoft/yii2/discussions)?

**Comment by samdark** [comment]
Converted it to discussion.
